### PR TITLE
"LGTM!"

### LINKED ISSUE #14: Requesting information about recording
The Recorder class takes the current game state and serializes it. How do you plan on using the information obtained from it? (I assume for training?)

It would be useful to know the purpose of this class, as well as the design process behind it, so we can maybe add more fields to the frame if they are relevant.

### LINKED ISSUE #29: More project refactoring
- [x] The current pathfinding system is all over the place. At the moment, `PathfindingHandler`, `MovementHandler` and `TasksHandler` all touch movement/pathfinding fields. These should be refactored.
- [x] Currently tasks are being opened using heinous methods, we should instead merge `MinigameHandler` and `TasksHandler` and allow each task to specify when it should be opened using something like MinigameSolver in #26 
- [x] `VisionHandler` is a mess, it should be refactored to reduce duplicated code, reduce method complexity, and stuff like that
- [x] The "Handler" singleton system might not be the best, some of the classes can become static, while others should have one instance per game and be reset once the game ends.
- [x] if the recording is meant to be separately given out to people, maybe it should be moved to it's own project or at least namespace

### LINKED ISSUE #36: Multi-thread pathfinding
We should look into running the pathfinding on (one or more) different threads so its more efficient and can be used to pathfind to all 10 available tasks at once

### PR DESCRIPTION
**Pull request by Alexejhero**
_Vedal let me out please it's cold down here_

---

### Changes made in this pull request
- Use Protobuf instead of JSON for serialization
- Make `PathfindingHandler` a singleton and run it on separate thread
- Create unstable collections, each element has a UnityEngine.Object anchor, if that is destroyed then the element is automatically removed from the collection
- Split `VisionHandler` into `OtherPlayersRecorder` and `DeadBodiesRecorder` monobehaviours
- Split `ImpostorHandler` into `MapDataRecorder` and other excluded class
- Add Event system for `GameStarted`, `MeetingStarted` and `MeetingEnded` events so we dont need 100 patches for the same thing
- Create `InteractionsHandler`
- Record task data for all tasks + sabotage
- Make socket connection resilient to disconnects and run on separate thread
- Other random stuff, pretty much a rewrite of 70% of the codebase, this list isn't even exhaustive

---

❗ Neural network broken, pls fix

---

Co-authored by @alekso56

Closes #14
Closes #29
Closes #36

### CODE REVIEW

**Comment by alekso56** [comment]
When socket is not connected on startup, the player is forced to walk down forever.

**Comment by Alexejhero** [comment]
> When socket is not connected on startup, the player is forced to walk down forever.

That was on purpose to test the movement handler, but it's been removed since.

The python side has not been updated to integrate with the new socket serialization system anyway.

**Comment by alekso56** [comment]
Im curious, if you are going to send partial strings anyways, why not make a data class and then send that as json then decode as json on python side?

**Comment by Alexejhero** [comment]
> Im curious, if you are going to send partial strings anyways, why not make a data class and then send that as json then decode as json on python side?

JSON has a lot of fluff (for example the name of the keys is repeated in every object) so it's very inefficient for socket purposes. Additionally, since we want to be saving this data to a file, we want the file to be as small as possible.

**Comment by alekso56** [comment]
maybe can do something like this on python side, com protocol is currently sending zero data from the client, so might want to fix that first.
```

from dataclasses import dataclass

from numpy import byte

from bytechomp import Reader


@dataclass
class PositionData:
    TotalDistance: float
    NextNodePositionX: float
    NextNodePositionY: float


@dataclass
class DeadBodyData:
    ParentId: byte
    LastSeenPositionX: float
    LastSeenPositionY: float
    FirstSeenTime: float
    NearbyPlayersCount: byte
    NearbyPlayers: list[byte]


@dataclass
class DoorData:
    position: PositionData
    IsOpen: bool


@dataclass
class VentData:
    position: PositionData
    ConnectingVentsCount: byte
    ConnectingVents: list[PositionData]


@dataclass
class OtherPlayerData:
    id: byte
    LastSeenPositionX: float
    LastSeenPositionY: float
    LastSeenTime: float
    TimesSawVent: byte
    RoundTimeVisible: float
    GameTimeVisible: float


@dataclass
class Frame:
    seenBodiesCount: byte
    seenBodies: list[DeadBodyData]
    DidKill: bool
    SabotageUsed: byte
    DoorsUsed: byte
    DidReport: bool
    DidVent: bool
    NearbyDoorsCount: byte
    NearbyDoors: list[DoorData]
    NearbyVentsCount: byte
    NearbyVents: list[VentData]
    LastSeenCount: byte
    LastSeen: list[OtherPlayerData]


def read(stream):
    print("Creating frame from byte array")
    reader = Reader[Frame]().allocate()
    if reader << stream.read(512):
        # construct dataclass
        my_struct = reader.build()
        reader.clear()
        return my_struct
```

**Review by Vedal987 — CHANGES_REQUESTED** [review]
do better serialization

**Comment by Alexejhero** [comment]
This PR will be ready for a re-riew once I've added deserialization on the python side

File: Neuro/Recording/Frame.proto [hunk 1]
```diff
@@ -2,12 +2,14 @@
...
 
 import "Recording/DeadBodies/DeadBodiesFrame.proto";
+import "Recording/Header/HeaderFrame.proto";
 import "Recording/LocalPlayer/LocalPlayerFrame.proto";
 import "Recording/Map/MapFrame.proto";
 import "Recording/OtherPlayers/OtherPlayersFrame.proto";
 
 message Frame {
     DeadBodiesFrame DeadBodies = 1;
+    optional HeaderFrame Header = 5;
```
Comment (JohnyDaison) [inline Neuro/Recording/Frame.proto]: Why did you put it on this line and not the end the block?

**Comment by Vedal987** [comment]
idk probably fine

**Review by Vedal987 — APPROVED** [review]

**Pull request merged**